src/biweight.jl

```diff
@@ -0,0 +1,179 @@
+```math
+\\begin{aligned}
+\\mathrm{MAD}_x &= \\mathrm{median}(\\left\\{\\left|x_i - M_x\\right|\\right\\}) \\\\
+\\mathrm{MAD}_y &= \\mathrm{median}(\\left\\{\\left|y_i - M_y\\right|\\right\\}).
+\\end{aligned}
+```
+
+If either ``\\mathrm{MAD}_x = 0`` or ``\\mathrm{MAD}_y = 0``, the pairwise covariance is
+defined to be zero.
+
+The parameter ``c`` is a tuning constant, for which the default is ``9.0``.
+Larger values will reduce the number of outliers that are removed — i.e. reducing
+robustness, but increasing sample efficiency.
+
+# Fields
+- `c::Float64`: The tuning constant corresponding to ``c`` above.
+- `modify_sample_size::Bool`: If `false`, then we use a sample size ``n_s`` equal to the
+    total number of observations ``n``. This is consistent with the standard definition of
+    biweight midcovariance in the literature. Otherwise, we count only those elements which
+    are not rejected as outliers in the numerator, i.e. those for which ``|u_i|<1``
+    and ``|v_i|<1``.
+    This follows the implementation in astropy [2].
+
+# Complexity
+- Space: ``O(p^2)``
+- Time: ``O(np^2)``
+
+# References
+[1] Beers, Flynn, and Gebhardt (1990; AJ 100, 32)
+"Measures of Location and Scale for Velocities in Clusters of Galaxies -- A Robust Approach"
+
+[2] [Astropy biweight_midcovariance](https://docs.astropy.org/en/stable/api/astropy.stats.biweight_midcovariance.html)
+
+[3] [NIST Dataplot biweight midcovariance](https://www.itl.nist.gov/div898/software/dataplot/refman2/auxillar/biwmidc.htm)
+"""
+struct BiweightMidcovariance <: CovarianceEstimator
+    c::Float64
+    modify_sample_size::Bool
+end
+function BiweightMidcovariance(; c::Float64=9.0, modify_sample_size::Bool=false)
+    return BiweightMidcovariance(c, modify_sample_size)
+end
+
+function covzm(ce::BiweightMidcovariance, x::AbstractVector{<:Real})
+    MADx = median(abs.(x))
+    numerator = zero(eltype(x))
+
+    # If MADx is zero, return zero.
+    MADx == 0 && return zero(one(eltype(x)) / 1)
+
+    denominator = zero(eltype(x))
+    count = 0
+    for xi in x
+        ui² = (xi / (ce.c * MADx))^2
+        ui² >= 1 && continue
+        count += 1
+        numerator += xi^2 * (1 - ui²)^4
+        denominator += (1 - ui²) * (1 - 5 * ui²)
+    end
+    n = ce.modify_sample_size ? count : length(x)
+    return n * numerator / (denominator^2)
+end
+
+function cov(
+    ce::BiweightMidcovariance,
+    x::AbstractVector{<:Real};
+    mean::Union{Nothing,<:Real}=nothing,
+)
+    Mx = isnothing(mean) ? median(x) : mean
+    return covzm(ce, x .- Mx)
+end
+
+function covzm(
+    ce::BiweightMidcovariance, x::AbstractVector{<:Real}, y::AbstractVector{<:Real}
+)
+    MADx = median(abs.(x))
+    MADy = median(abs.(y))
+
+    # Promote types between x & y for numerator
+    numerator = zero(promote_type(eltype(x), eltype(y)))
+
+    # If either of the MADs are zero, return zero.
+    MADx == 0 && return numerator
+    MADy == 0 && return numerator
```

Comment:
same comment as earlier, also you write here `return 0` but you return the `numerator`?

Reply:
`numerator` here is zero, but of the correct output type (which could be different from `eltype(x)` - see the line above). [Changing to use `iszero` though]

Reply:
(I expanded the comment with this justification.)